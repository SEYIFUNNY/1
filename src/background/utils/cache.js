import initCache from '#/common/cache';
import { commands } from './message';

const cache = initCache({
  lifetime: 5 * 60 * 1000,
});

Object.assign(commands, {
  CacheLoad(data) {
    return cache.get(data) || null;
  },
  CacheHit(data) {
    cache.hit(data.key, data.lifetime);
  },
});

export default cache;
