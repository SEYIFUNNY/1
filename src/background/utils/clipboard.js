let clipboardData;
function onCopy(e) {
  document.removeEventListener("copy", onCopy, true);
  e.stopImmediatePropagation();
  e.preventDefault();
  const { type, data } = clipboardData;
  e.clipboardData.setData(type || 'text/plain', data);
}
export default function setClipboard(data) {
  clipboardData = data;
  document.addEventListener("copy", onCopy, true);
  const ret = document.execCommand('copy', false, null);
  if (!ret && process.env.DEBUG) {
    console.warn('Copy failed!');
  }
}
