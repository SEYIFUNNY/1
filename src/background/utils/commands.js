import { getActiveTab } from '#/common';
import { tabOpen } from './tabs';

const ROUTES = {
  newScript: '#scripts/_new',
  settings: '#settings',
};

global.addEventListener('backgroundInitialized', () => {
  browser.commands.onCommand.addListener(async (cmd) => {
    const tab = await getActiveTab();
    const optionsUrl = browser.runtime.getURL(browser.runtime.getManifest().options_ui.page);
    const url = `${optionsUrl}${ROUTES[cmd] || ''}`;
    tabOpen({ url, insert: true }, { tab });
  });
}, { once: true });
