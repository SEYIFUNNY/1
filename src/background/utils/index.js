export { default as cache } from './cache';
export { default as getEventEmitter } from './events';
export { initialize } from './init';
export * from './message';
export * from './options';
export * from './search';
