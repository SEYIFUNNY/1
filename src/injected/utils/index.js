import { CustomEvent, jsonDump, jsonLoad } from './helpers';

export { sendMessage, request, throttle } from 'src/common';

export function postData(destId, data) {
  // Firefox issue: data must be stringified to avoid cross-origin problem
  const e = new CustomEvent(destId, { detail: jsonDump(data) });
  document.dispatchEvent(e);
}

let doInject;
export function inject(code) {
  if (!doInject) {
    const id = getUniqId('VM-');
    const detect = domId => {
      const span = document.createElement('span');
      span.id = domId;
      document.documentElement.appendChild(span);
    };
    injectViaText(`(${detect.toString()})(${jsonDump(id)})`);
    const span = document.querySelector(`#${id}`);
    if (span) {
      span.parentNode.removeChild(span);
      doInject = injectViaText;
    } else {
      // For Firefox in CSP limited pages
      doInject = injectViaBlob;
    }
  }
  doInject(code);
}

function injectViaText(code) {
  const script = document.createElement('script');
  const doc = document.body || document.documentElement;
  script.textContent = code;
  doc.appendChild(script);
  try {
    doc.removeChild(script);
  } catch (e) {
    // ignore if body is changed and script is detached
  }
}

// Firefox does not support script injection by `textCode` in CSP limited pages
// have to inject via blob URL, leading to delayed first injection
function injectViaBlob(code) {
  const script = document.createElement('script');
  const doc = document.body || document.documentElement;
  // https://en.wikipedia.org/wiki/Byte_order_mark
  const blob = new Blob(['\ufeff', code], { type: 'text/javascript' });
  const url = URL.createObjectURL(blob);
  script.src = url;
  doc.appendChild(script);
  try {
    doc.removeChild(script);
  } catch (e) {
    // ignore if body is changed and script is detached
  }
  URL.revokeObjectURL(url);
}

export function getUniqId() {
  return Date.now().toString(36) + Math.random().toString(36).slice(2, 6);
}

export function bindEvents(srcId, destId, handle) {
  document.addEventListener(srcId, e => {
    const data = jsonLoad(e.detail);
    handle(data);
  }, false);
  return data => { postData(destId, data); };
}

export function attachFunction(id, cb) {
  Object.defineProperty(window, id, {
    value(...args) {
      cb.apply(this, args);
      delete window[id];
    },
    configurable: true,
  });
}
