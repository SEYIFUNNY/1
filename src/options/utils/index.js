import { reactive } from 'vue';
import { sendCmdDirectly } from '@/common';
import { route } from '@/common/router';

export * from './search';

export const store = reactive({
  route,
  /** Speedup and deflicker initial page load by not rendering an invisible script list */
  canRenderScripts: [SCRIPTS, TAB_RECYCLE, ''].includes(route.hash),
  scripts: [],
  removedScripts: [],
  importing: null,
  loading: false,
  /** Whether removed scripts need to be filtered from `store.scripts`. */
  needRefresh: false,
  storageSize: 0,
  sync: [],
  title: null,
});

export function markRemove(script, removed) {
  return sendCmdDirectly('MarkRemoved', {
    id: script.props.id,
    removed,
  });
}
