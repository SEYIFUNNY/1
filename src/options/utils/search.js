export function createSearchRules(search) {
  const tagLists = [[], []];
  /**
   * @type Array<{
   *   scope: string;
   *   pattern: string | RegExp;
   *   negative: boolean;
   * }>
   */
  const rules = [];
  search = search.toLowerCase();
  let offset = 0;
  while (search[offset] === ' ') offset += 1;
  while (offset < search.length) {
    const negative = search[offset] === '!';
    if (negative) offset += 1;
    let prefix =
      search.slice(offset).match(/^(#|re:|(?:name|code)(?:\+re)?:)/)?.[1] || '';
    if (prefix) offset += prefix.length;
    const quote =
      (!prefix || prefix.endsWith(':')) &&
      search.slice(offset).match(/^['"]/)?.[0];
    if (quote) offset += 1;
    let pattern = '';
    const endChar = quote || ' ';
    while (offset < search.length) {
      const ch = search[offset];
      if (quote && ch === quote && search[offset + 1] === quote) {
        // escape quotes by double it
        pattern += quote;
        offset += 2;
      } else if (ch !== endChar) {
        pattern += ch;
        offset += 1;
      } else {
        break;
      }
    }
    if (quote) {
      if (offset < search.length) offset += 1;
      else throw new Error('Unmatched quotes');
    }
    if (prefix === '#') {
      tagLists[+negative].push(pattern);
    } else {
      // Strip ':'
      prefix = prefix.slice(0, -1);
      if (/(?:^|\+)re$/.test(prefix)) {
        prefix = prefix.slice(0, -3);
        rules.push({
          scope: prefix,
          pattern: new RegExp(pattern, 'i'),
          negative,
        });
      } else {
        const reMatches = pattern.match(/^\/(.*?)\/(\w*)$/);
        if (reMatches) pattern = new RegExp(reMatches[1], reMatches[2] || 'i');
        rules.push({
          scope: prefix,
          pattern,
          negative,
        });
      }
    }
    while (search[offset] === ' ') offset += 1;
  }
  for (let negative = 0; negative < 2; negative += 1) {
    const tags = tagLists[negative];
    if (tags.length) {
      const sanitizedTags = tags
        .map((tag) => tag.replace(/[^\w.-]/g, '').replace(/\./g, '\\.'))
        .filter(Boolean)
        .join('|');
      rules.unshift({
        scope: 'tags',
        pattern: new RegExp(`(?:^|\\s)(${sanitizedTags})(\\s|$)`),
        negative: !!negative,
      });
    }
  }
  return rules;
}

export function testSearchRule(rule, data) {
  const { pattern, negative } = rule;
  const result =
    typeof pattern.test === 'function'
      ? pattern.test(data)
      : data.includes(pattern);
  return negative ^ result;
}
