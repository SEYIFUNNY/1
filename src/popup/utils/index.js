export const store = {
  scripts: [],
  frameScripts: [],
  commands: [],
  domain: '',
  injectionFailure: null,
  injectable: true,
  blacklisted: false,
};
