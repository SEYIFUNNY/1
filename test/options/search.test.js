import { createSearchRules } from '@/options/utils/search';

test('createSearchRules', () => {
  expect(createSearchRules('')).toMatchSnapshot();
  expect(createSearchRules('#a #b !#c hello')).toMatchSnapshot();
  expect(createSearchRules('#a-b #b name:hello world')).toMatchSnapshot();
  expect(createSearchRules('#a.b #b name:"hello world"')).toMatchSnapshot();
  expect(createSearchRules('#a.b #b name+re:"hello world"')).toMatchSnapshot();
  expect(createSearchRules('#a.b #b !name+re:"hello world"')).toMatchSnapshot();
  expect(createSearchRules('"#a.b" !"#b"')).toMatchSnapshot();
  expect(createSearchRules('/regexp/ code:/regexp/u /not regexp/')).toMatchSnapshot();
  expect(createSearchRules('foobar re:foobar name+re:foobar code+re:foobar')).toMatchSnapshot();
});
