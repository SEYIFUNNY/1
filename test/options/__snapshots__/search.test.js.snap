// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createSearchRules 1`] = `[]`;

exports[`createSearchRules 2`] = `
[
  {
    "negative": true,
    "pattern": /\\(\\?:\\^\\|\\\\s\\)\\(c\\)\\(\\\\s\\|\\$\\)/,
    "scope": "tags",
  },
  {
    "negative": false,
    "pattern": /\\(\\?:\\^\\|\\\\s\\)\\(a\\|b\\)\\(\\\\s\\|\\$\\)/,
    "scope": "tags",
  },
  {
    "negative": false,
    "pattern": "hello",
    "scope": "",
  },
]
`;

exports[`createSearchRules 3`] = `
[
  {
    "negative": false,
    "pattern": /\\(\\?:\\^\\|\\\\s\\)\\(a-b\\|b\\)\\(\\\\s\\|\\$\\)/,
    "scope": "tags",
  },
  {
    "negative": false,
    "pattern": "hello",
    "scope": "name",
  },
  {
    "negative": false,
    "pattern": "world",
    "scope": "",
  },
]
`;

exports[`createSearchRules 4`] = `
[
  {
    "negative": false,
    "pattern": /\\(\\?:\\^\\|\\\\s\\)\\(a\\\\\\.b\\|b\\)\\(\\\\s\\|\\$\\)/,
    "scope": "tags",
  },
  {
    "negative": false,
    "pattern": "hello world",
    "scope": "name",
  },
]
`;

exports[`createSearchRules 5`] = `
[
  {
    "negative": false,
    "pattern": /\\(\\?:\\^\\|\\\\s\\)\\(a\\\\\\.b\\|b\\)\\(\\\\s\\|\\$\\)/,
    "scope": "tags",
  },
  {
    "negative": false,
    "pattern": /hello world/i,
    "scope": "name",
  },
]
`;

exports[`createSearchRules 6`] = `
[
  {
    "negative": false,
    "pattern": /\\(\\?:\\^\\|\\\\s\\)\\(a\\\\\\.b\\|b\\)\\(\\\\s\\|\\$\\)/,
    "scope": "tags",
  },
  {
    "negative": true,
    "pattern": /hello world/i,
    "scope": "name",
  },
]
`;

exports[`createSearchRules 7`] = `
[
  {
    "negative": false,
    "pattern": "#a.b",
    "scope": "",
  },
  {
    "negative": true,
    "pattern": "#b",
    "scope": "",
  },
]
`;

exports[`createSearchRules 8`] = `
[
  {
    "negative": false,
    "pattern": /regexp/i,
    "scope": "",
  },
  {
    "negative": false,
    "pattern": /regexp/u,
    "scope": "code",
  },
  {
    "negative": false,
    "pattern": "/not",
    "scope": "",
  },
  {
    "negative": false,
    "pattern": "regexp/",
    "scope": "",
  },
]
`;

exports[`createSearchRules 9`] = `
[
  {
    "negative": false,
    "pattern": "foobar",
    "scope": "",
  },
  {
    "negative": false,
    "pattern": /foobar/i,
    "scope": "",
  },
  {
    "negative": false,
    "pattern": /foobar/i,
    "scope": "name",
  },
  {
    "negative": false,
    "pattern": /foobar/i,
    "scope": "code",
  },
]
`;
